cpp/src/io/parquet/writer_impl.cu

```diff
@@ -70,364 +73,515 @@ parquet::Compression to_parquet_compression(compression_type compression)
-      _def_level              = std::move(dremel.def_level);
-      leaf_col_offset         = dremel.leaf_col_offset;
-      _data_count             = dremel.leaf_data_size;
-      _max_def_level          = dremel.max_def_level;
-
-      _type_width = (is_fixed_width(_leaf_col.type())) ? cudf::size_of(_leaf_col.type()) : 0;
-      _data       = (is_fixed_width(_leaf_col.type()))
-                ? _leaf_col.head<uint8_t>() + leaf_col_offset * _type_width
-                : nullptr;
-
-      // Calculate nesting levels
-      column_view curr_col = col;
-      _nesting_levels      = 0;
-      while (curr_col.type().id() == type_id::LIST) {
-        lists_column_view list_col(curr_col);
-        _nesting_levels++;
-        curr_col = list_col.child();
-      }
-
-      // Update level nullability if no nullability was passed in.
-      curr_col = col;
-      if (_nullability.empty()) {
-        while (curr_col.type().id() == type_id::LIST) {
-          lists_column_view list_col(curr_col);
-          _nullability.push_back(list_col.null_mask() != nullptr);
-          curr_col = list_col.child();
-        }
-        _nullability.push_back(curr_col.null_mask() != nullptr);
-      }
+      gpu::dremel_data dremel =
+        gpu::get_dremel_data(cudf_col, _d_nullability, _nullability, stream);
+      _dremel_offsets = std::move(dremel.dremel_offsets);
+      _rep_level      = std::move(dremel.rep_level);
+      _def_level      = std::move(dremel.def_level);
+      _data_count = dremel.leaf_data_size;  // Needed for knowing what size dictionary to allocate
 
       stream.synchronize();
     } else {
-      if (_nullability.empty()) { _nullability = {col.nullable()}; }
-      _max_def_level = (_nullability[0]) ? 1 : 0;
+      // For non-list struct, the size of the root column is the same as the size of the leaf column
+      _data_count = cudf_col.size();
+    }
+  }
+
+  column_view cudf_column_view() const { return cudf_col; }
+  parquet::Type physical_type() const { return schema_node.type; }
+
+  column_view leaf_column() const
+  {
+    auto col = cudf_col;
+    while (cudf::is_nested(col.type())) {
+      if (col.type().id() == type_id::LIST) {
+        col = col.child(lists_column_view::child_column_index);
+      } else if (col.type().id() == type_id::STRUCT) {
+        col = col.child(0);  // Stored cudf_col has only one child if struct
+      }
     }
-    if (_string_type && _data_count > 0) {
-      strings_column_view view{_leaf_col};
-      _indexes = rmm::device_buffer(_data_count * sizeof(gpu::nvstrdesc_s), stream);
-
-      stringdata_to_nvstrdesc<<<((_data_count - 1) >> 8) + 1, 256, 0, stream.value()>>>(
-        reinterpret_cast<gpu::nvstrdesc_s *>(_indexes.data()),
-        view.offsets().data<size_type>() + leaf_col_offset,
-        view.chars().data<char>(),
-        _nulls,
-        _data_count);
-      _data = _indexes.data();
+    return col;
+  }
 
-      stream.synchronize();
+  gpu::EncColumnDesc get_device_view()
```

Comment:
Would get_column_descriptor or something else be a better name here? Without much context, the function name implies that a column_device_view is returned here.

Reply:
But it's a member function of `parquet_column_view`.

Reply:
You agree with **Kumar Aatish's** comment 🙂

Reply:
Lol, how did I read the wrong name?